### PR TITLE
drivers: crypto: stm32_saes: fallback to software on 192bit AES keys

### PR DESCRIPTION
**Pull request by etienne-lms**
Implement AES software operation for when AES key is 192 bit that is not supported by the STM32 SAES engine.
This P-R includes cleanup and preparatory patches followed by the above mentioned AES-192b software fallback implementation.

### CODE REVIEW

**Comment by etienne-lms** [comment]
@tbourgoi, could you have a look at this series?

**Comment by jforissier** [comment]
> @jbech-linaro, this series is not related to changes related to IBART. Any idea why the test failed?

You need to rebase to include fcabe15c7 :wink:

File: core/drivers/crypto/stm32/cipher.c [hunk 1]
```diff
@@ -225,12 +197,21 @@ static TEE_Result stm32_cryp_cipher_allocate(void **ctx, uint32_t algo)
 	}
 }
 
-/*
- * Allocate the SW cipher data context for SAES peripheral.
- *
```
Comment (tbourgoi) [inline core/drivers/crypto/stm32/cipher.c:230]: why did you remove the documentation ?

**Review by tbourgoi — COMMENTED** [review]
I think the word "one" can be removed in the commit message : 
"to place them next one to each".

File: core/drivers/crypto/stm32/cipher.c (hunk 1 above)
Comment (etienne-lms) [reply core/drivers/crypto/stm32/cipher.c:230]: i think it is not useful. These are a local function and then label is already quite explicit.

**Comment by etienne-lms** [comment]
@tbourgoi, do you want me to update that commit message before you review these changes?

**Review by tbourgoi — COMMENTED** [review]
First sentence of the commit looks strange to me. I suggest : 
`Implement` AES software operation for 192 bits key that is
not supported by the STM32 SAES peripheral.`

**Comment by tbourgoi** [comment]
> @tbourgoi, do you want me to update that commit message before you review these changes?

Yes please, apart from the 2 commit messages I mentioned others looks okay to me.
I'll review one last time when you have update the messages

**Comment by etienne-lms** [comment]
I've updated the series, fixed the last commit (@jforissier, it's generic, not stm32 specific, maybe you want to ack it) and rebased for CI test purpose.

**Comment by jbech-linaro** [comment]
> > @jbech-linaro, this series is not related to changes related to IBART. Any idea why the test failed?
> 
> You need to rebase to include [fcabe15](https://github.com/OP-TEE/optee_os/commit/fcabe15c7783f14d6997a89154a8754790c648ea) 😉 (remember that IBART builds the PR branch without merging into master, i.e., the checked out code is pull/<PR_ID>/head not pull/<PR_ID>/merge)

Noticed that a lot of stuff has failed the last few days in IBART, I'll have a look.

**Comment by jbech-linaro** [comment]
IBART: I think I've sorted it now. It was a combination of permission issues and that we needed to prime a ccache (longer initial build time), which probably meant timeouts. I'll restart the jobs that hasn't been merged yet.

**Comment by etienne-lms** [comment]
Thanks @jbech-linaro.
Hi @tbourgoi, I think this P-R is in a sufficiently good state for a review.

**Comment by tbourgoi** [comment]
Hi @etienne-lms I saw a discussion about software fallback https://github.com/OP-TEE/optee_docs/pull/235
Maybe It could be usefull for us here.
Something like CFG_STM32_SAES_SW_FALLBACK (I open to suggestion for the name of the configuration switch)
For me, it should be enabled by default in the platform configuration.

File: core/drivers/crypto/stm32/cipher.c (hunk 1 above)
Comment (tbourgoi) [reply core/drivers/crypto/stm32/cipher.c:230]: ok, valid point

**Comment by etienne-lms** [comment]
You're right. I'll add such a switch. `CFG_STM32_SAES_SW_FALLBACK=y|n` sounds ok to me.
Thanks

**Comment by etienne-lms** [comment]
Comment addressed.

File: core/drivers/crypto/stm32/cipher.c [hunk 2]
```diff
@@ -119,6 +119,12 @@ static TEE_Result saes_init(union ip_ctx *ip_ctx, bool is_decrypt,
 		TEE_OperationMode mode = TEE_MODE_ILLEGAL_VALUE;
 		TEE_Result res = TEE_ERROR_GENERIC;
 
+		if (!IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK)) {
+			DMSG("STM32 SAES does not support 192bit keys");
```
Comment (tbourgoi) [inline core/drivers/crypto/stm32/cipher.c:123]: I'm thinking if EMSG is more relevant here or not ? 
It's an error so error message, but maybe EMSG will make it too verbose and we can assume the user will recompile the driver with debug log level to ease debug.

**Review by tbourgoi — COMMENTED** [review]
you can add my tag on 58f752628f83868a3c2cf6d33b70a58f33a29096 : 
`Reviewed-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>`

File: core/drivers/crypto/stm32/cipher.c [hunk 3]
```diff
@@ -177,8 +187,11 @@ static void saes_copy_state(union ip_ctx *dst_ip_ctx, union ip_ctx *src_ip_ctx)
 	dst_ip_ctx->saes.algo = src_ctx->algo;
 	dst_ip_ctx->saes.use_fallback = src_ctx->use_fallback;
 
-	if (src_ctx->use_fallback)
+	if (IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK) && src_ctx->use_fallback) {
+		assert(IS_ENABLED(CFG_STM32_SAES_SW_FALLBACK));
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/crypto/stm32/cipher.c:191]: it is! my bad. i'll fix.

File: core/drivers/crypto/stm32/cipher.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/crypto/stm32/cipher.c:123]: Makes sense. I'll switch to `EMSG()`.

**Comment by etienne-lms** [comment]
Comments addressed.
@tbourgoi, may I squash the fixup commits?

**Comment by tbourgoi** [comment]
> Comments addressed. @tbourgoi, may I squash the fixup commits?

Yes please

**Comment by tbourgoi** [comment]
All commits look good, you can add my tag on the 5 commits : 
`Reviewed-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>`

**Comment by etienne-lms** [comment]
trustedfirmware repositories seem a bit unstable. I've seen a few CI test errors reported with trace messages like the below, but it does not seem systematic:
```
fatal: unable to access 'https://git.trustedfirmware.org/hafnium/third_party/dtc/': server certificate verification failed. CAfile: none CRLfile: none
fatal: clone of 'https://git.trustedfirmware.org/hafnium/third_party/dtc' into submodule path '/__w/optee_os/optee_repo_qemu_v8/hafnium/third_party/dtc' failed
Failed to clone 'third_party/dtc'. Retry scheduled
